### PR TITLE
Abilityrework

### PR DESCRIPTION
**Pull request by hunharibo**
Attribute (data container) + ability rework in one PR. 

Only the skeleton of stuff is in here, some major functionality is not implemented (like miscastchange and effects for example or channeling cast type etc.) but the templating is in place and some proof of concept implementations too.

Not necessarily meant to be reviewed "officially" for merge, but it would be nice if you all could look at it.

### CODE REVIEW

File: CSharpSourceCode/ObjectDataExtensions/AgentExtendedInfoComponent.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
+﻿using TaleWorlds.MountAndBlade;
+using TOW_Core.Battle.Extensions;
+
+namespace TOW_Core.AttributeDataSystem
+{
+    /// <summary>
+    /// Static Attribute attached to an agent on Battlefield
```
Comment (Z3rca) [inline CSharpSourceCode/ObjectDataExtensions/AgentExtendedInfoComponent.cs]: Change text here

Comment (Z3rca) [reply CSharpSourceCode/ObjectDataExtensions/AgentExtendedInfoComponent.cs]: removing and copypasting my own script 😛

File: CSharpSourceCode/Abilities/Ability.cs [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+                        {
+                            ActivateAbility(casterAgent);
+                        }
+                    };
+                    timer.Start();
+                }
+            }
+        }
+
+        public virtual void ActivateAbility(Agent casterAgent)
```
Comment (Z3rca) [inline CSharpSourceCode/Abilities/Ability.cs:92]: you can put this logic on Triggered Effect System, reading out the provided template and using method acting on the template data. with that, we can reuse the same logic for Grenade effects, or even magic melee attack effects.

File: CSharpSourceCode/Abilities/AbilityFactory.cs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using System.Xml.Serialization;
+using TaleWorlds.Library;
+
+namespace TOW_Core.Abilities
+{
+    public class AbilityFactory
```
Comment (Z3rca) [inline CSharpSourceCode/Abilities/AbilityFactory.cs:12]: Even I like Factory, I dont know if this is a helpful name in later stages... Have to think of a name... Like AbilityTemplateAssemblyManager? I am  not good in this... so this is really just a minor :D

File: CSharpSourceCode/Abilities/AbilityTemplate.cs [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        public string Name = "";
+        [XmlAttribute]
+        public string SpriteName  = "";
+        [XmlAttribute]
+        public int CoolDown = 10;
+        [XmlAttribute]
+        public int WindsOfMagicCost = 0; //spell only
+        [XmlAttribute]
+        public float BaseMisCastChance = 0.3f; //spell only
+        [XmlAttribute]
```
Comment (Z3rca) [inline CSharpSourceCode/Abilities/AbilityTemplate.cs:27]: up to here I would leave it in the Ability Template, with a Reference to a TriggeredEffectTemplateID, The rest goes into a TriggeredEffectTemplate

File: CSharpSourceCode/Battle/TriggeredEffect/TriggeredEffectTemplate.cs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        [XmlAttribute]
+        public float SoundEffectLength = 2.5f;
+        [XmlAttribute]
+        public DamageType DamageType = DamageType.Fire;
+        [XmlAttribute]
+        public int DamageAmount = 50;
+        [XmlAttribute]
+        public float Radius = 5;
+        [XmlAttribute]
+        public bool FriendlyFire = true;
```
Comment (Z3rca) [inline CSharpSourceCode/Battle/TriggeredEffect/TriggeredEffectTemplate.cs]: add also a "FriendsOnly"  bool, or we have to work with flags here... e.g. Magic healing could only affect friendly targets

File: CSharpSourceCode/ObjectDataExtensions/ExtendedInfoMissionLogic.cs [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+            base.OnMissionTick(dt);
+            foreach(var agent in Mission.Agents)
+            {
+                var hero = agent.GetHero();
+                if (hero != null && hero.IsSpellCaster())
+                {
+                    var info = hero.GetExtendedInfo();
+                    if(info != null)
+                    {
+                        info.CurrentWindsOfMagic += dt;
```
Comment (Z3rca) [inline CSharpSourceCode/ObjectDataExtensions/ExtendedInfoMissionLogic.cs]: balancing wise, that shouldn't be but this is really just a minor and open for discussion :D

**Review by Z3rca — COMMENTED** [review]
In general, very nice work, you did improve the system, and already provided additional functions.

My concerns currently are just the TriggeredEffects, Spells. I believe without much reshuffling, we can at this point make the system dynamic enough, to allow also Weapon effects (grenades or even melee weapons) or things like traps or Map specific effects (like a lava floor).   I would like if we can meet and talk this through. I know I am hard headed, and it starts to get already complicated, but we can achieve a lot with that. Beside that, very good work! :)

File: CSharpSourceCode/Battle/TriggeredEffect/TriggeredEffect.cs [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        {
+            _template = template;
+        }
+
+        public void Dispose()
+        {
+            CleanUp();
+        }
+
+        public void Trigger(Vec3 position, Vec3 normal, Agent triggererAgent)
```
Comment (Z3rca) [inline CSharpSourceCode/Battle/TriggeredEffect/TriggeredEffect.cs:21]: maybe also make an overload for a single agent target something Like Trigger(Agent target, Agent triggererAgent)

File: CSharpSourceCode/ObjectDataExtensions/ExtendedInfoMissionLogic.cs (hunk 6 above)
Comment (hunharibo) [reply CSharpSourceCode/ObjectDataExtensions/ExtendedInfoMissionLogic.cs]: yeah, i just kept your original implementation haha :D the designers can tell us the rate at which it should recharge.

File: CSharpSourceCode/Battle/TriggeredEffect/TriggeredEffectTemplate.cs (hunk 5 above)
Comment (hunharibo) [reply CSharpSourceCode/Battle/TriggeredEffect/TriggeredEffectTemplate.cs]: TODO, replace this with enum targetType

File: CSharpSourceCode/ObjectDataExtensions/ExtendedInfoMissionLogic.cs (hunk 6 above)
Comment (hunharibo) [reply CSharpSourceCode/ObjectDataExtensions/ExtendedInfoMissionLogic.cs]: TODO: introduce hero level recharge rate modifier

**Review by Z3rca — APPROVED** [review]

**Review by martkartasev — APPROVED** [review]

**Pull request merged**